File: server/modules/home/home.controller.ts
import { Controller, Get, Query, Req, UseGuards } from '@nestjs/common';
import { HomeService } from './home.service';
import type { Request } from 'express';
import { SessionUserGuard } from '../../guards/session-user.guard';

function parseYm(year?: string, month?: string): { year?: number; month?: number } {
  if (year == null || month == null) return {};
  const y = parseInt(year, 10);
  const m = parseInt(month, 10);
  if (!Number.isFinite(y) || !Number.isFinite(m)) return {};
  return { year: y, month: m };
}

@Controller('api/home')
@UseGuards(SessionUserGuard)
export class HomeController {
  constructor(private readonly homeService: HomeService) {}

  @Get('todos')
  async getTodos(
    @Req() req: Request,
    @Query('year') year?: string,
    @Query('month') month?: string,
  ) {
    return this.homeService.getTodos(req.userContext!.userId!, parseYm(year, month));
  }

  @Get('overview')
  async getOverview(
    @Req() req: Request,
    @Query('year') year?: string,
    @Query('month') month?: string,
  ) {
    return this.homeService.getOverview(req.userContext!.userId!, parseYm(year, month));
  }

  @Get('action-counts')
  async getActionCounts(
    @Req() req: Request,
    @Query('year') year?: string,
    @Query('month') month?: string,
  ) {
    return this.homeService.getActionCounts(req.userContext!.userId!, parseYm(year, month));
  }
}

File: server-java/src/main/java/com/jixiao2/server/home/HomeController.java
package com.jixiao2.server.home;

import com.jixiao2.server.web.CurrentUser;
import java.util.Map;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/home")
public class HomeController {

  private final HomeService homeService;

  public HomeController(HomeService homeService) {
    this.homeService = homeService;
  }

  @GetMapping("/todos")
  public Map<String, Object> todos(
      @CurrentUser String userId,
      @RequestParam(required = false) Integer year,
      @RequestParam(required = false) Integer month) {
    return homeService.getTodos(userId, year, month);
  }

  @GetMapping("/overview")
  public Map<String, Object> overview(
      @CurrentUser String userId,
      @RequestParam(required = false) Integer year,
      @RequestParam(required = false) Integer month) {
    return homeService.getOverview(userId, year, month);
  }

  @GetMapping("/action-counts")
  public Map<String, Object> actionCounts(
      @CurrentUser String userId,
      @RequestParam(required = false) Integer year,
      @RequestParam(required = false) Integer month) {
    return homeService.getActionCounts(userId, year, month);
  }
}
